src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -388,6 +388,7 @@ bool ObjectMonitor::enter(JavaThread* current) {
   { // Change java thread status to indicate blocked on monitor enter.
     JavaThreadBlockedOnMonitorEnterState jtbmes(current, this);
 
+    assert(current->current_pending_monitor() == NULL, "invariant");
```

Comment:
Really this belongs inside set_current_pending_monitor() where the only allowed transitions are null to non-null, and non-null to null. But that would be a separate RFE.